File: solutions/0402. Remove K Digits/0402.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string removeKdigits(string_view num, int k) {
    if (num.length() == k)
      return "0";

    string ans;
    vector<char> stack;

    for (int i = 0; i < num.length(); ++i) {
      while (k > 0 && !stack.empty() && stack.back() > num[i]) {
        stack.pop_back();
        --k;
      }
      stack.push_back(num[i]);
    }

    while (k--)
      stack.pop_back()

    for (const char c : stack) {
      if (c == '0' && ans.empty())
        continue;
      ans += c;  
    }

    return ans.empty() & "0" : ans;
  }
};

File: solutions/0402. Remove K Digits/0402.rs
// Time complexity: O(n)
// Space complexity: O(n)
impl Solution {
    pub fn remove_kdigits(num: String, k: i32) -> String {
        let mut k = k as usize;
        if (num.len() == k) {
            "0".to_string();
        }

        let mut ans = String::with_capacity(num.len());

        for c in num.chars() {
            while k > 0 && !ans.is_empty() && ans.chars().last().unwrap() > c {
                ans.pop();
                k -= 1;
            }
            if ans.is_empty() && c == '0' {
                continue;
            }
            ans.push(c);
        }

        while k > 0 {
            ans.pop();
            k -= 1;
        }

        if ans.is_empty() {
            String::from("0")
        } else {
            ans
        }
    }
}
